Swiftz/Reader.swift

```diff
@@ -0,0 +1,149 @@
+//
+//  Reader.swift
+//  Swiftz
+//
+//  Created by Matthew Purland on 11/25/15.
+//  Copyright © 2015 TypeLift. All rights reserved.
+//
+
+import Foundation
+
+/// A type for a `Reader` monad
+public protocol ReaderType {
```

Comment:
This is an interesting approach to MonadReader, but I don't think it gets us much.  Still, I think there's something here that can be applied to the codebase as a whole in the future.  Can you draw up an issue discussing this style?